coa/pkg/apis/v1alpha2/host/host.go

```diff
@@ -152,85 +165,123 @@ func (h *APIHost) Launch(config HostConfig,
 			return v1alpha2.NewCOAError(nil, fmt.Sprintf("no vendor factories can provide vendor type '%s'", v.Type), v1alpha2.BadConfig)
 		}
 	}
-	if len(h.Vendors) > 0 {
-		var evaluationContext *utils.EvaluationContext
-		for _, v := range h.Vendors {
-			if _, ok := v.Vendor.(vendors.IEvaluationContextVendor); ok {
-				log.Info("--- evaluation context established ---")
-				evaluationContext = v.Vendor.(vendors.IEvaluationContextVendor).GetEvaluationContext()
-			}
+	if len(h.Vendors) == 0 {
+		return v1alpha2.NewCOAError(nil, "no vendors are found", v1alpha2.MissingConfig)
+	}
+	var evaluationContext *utils.EvaluationContext
+	for _, v := range h.Vendors {
+		if _, ok := v.Vendor.(vendors.IEvaluationContextVendor); ok {
+			log.Info("--- evaluation context established ---")
+			evaluationContext = v.Vendor.(vendors.IEvaluationContextVendor).GetEvaluationContext()
 		}
+	}
 
-		if evaluationContext != nil {
-			for _, v := range h.Vendors {
-				log.Infof("--- evaluation context is sent to vendor: %s ---", v.Vendor.GetInfo().Name)
-				v.Vendor.SetEvaluationContext(evaluationContext)
-			}
-		}
-		var wg sync.WaitGroup
+	if evaluationContext != nil {
 		for _, v := range h.Vendors {
-			if v.LoopInterval > 0 {
-				if wait {
-					wg.Add(1)
-				}
-				go func(v VendorSpec) {
-					v.Vendor.RunLoop(time.Duration(v.LoopInterval) * time.Second)
-				}(v)
-			}
+			log.Infof("--- evaluation context is sent to vendor: %s ---", v.Vendor.GetInfo().Name)
+			v.Vendor.SetEvaluationContext(evaluationContext)
 		}
-		if len(config.Bindings) > 0 {
-			endpoints := make([]v1alpha2.Endpoint, 0)
-			for _, v := range h.Vendors {
-				endpoints = append(endpoints, v.Vendor.GetEndpoints()...)
+	}
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	for _, v := range h.Vendors {
+		if v.LoopInterval > 0 {
+			if wait {
+				wg.Add(1)
 			}
-
-			for _, b := range config.Bindings {
-				switch b.Type {
-				case "bindings.http":
+			go func(v VendorSpec) {
+				defer func() {
 					if wait {
-						wg.Add(1)
+						wg.Done()
 					}
-					var binding bindings.IBinding
-					var err error
-					if h.SharedPubSubProvider != nil {
-						binding, err = h.launchHTTP(b.Config, endpoints, h.SharedPubSubProvider.(pubsub.IPubSubProvider))
-					} else {
-						var bindingPubsub pv.IProvider
-						for _, providerFactory := range providerFactories {
-							mProvider, err := providerFactory.CreateProvider(
-								config.API.PubSub.Provider.Type,
-								config.API.PubSub.Provider.Config)
-							if err != nil {
-								return err
-							}
-							bindingPubsub = mProvider
-							break
+				}()
+				v.Vendor.RunLoop(ctx, time.Duration(v.LoopInterval)*time.Second)
```

Comment:
hello haishi, the vendor runloop is guaranteed to be called after vendor is initialized.  it's in line 151 to 155.

Reply:
The pub-sub provider could be initialized at line 217-223

Reply:
oh ok. I can adjust the sequence. Bindings can be moved before runloop. Sounds good?

Reply:
Hi @Haishi2016 , thanks for the comments. I still don't quite understand is in which case pubsub channel will be used before initialization. 

Please correct me if my understanding is not correct.
1. We will create and init all pubsub providers when creating vendors in main thread. Only when `config.API.PubSub.Shared = true`, `h.SharedPubSubProvider` will be assigned with a new pubsub provider which is already initialized by CreateProvider`, otherwise each vendor will use its own initialized pubsub.
2. Next, we will start new threads to call `vendor.RunLoop`, at this moment, even if `vendor.RunLoop` used the pubsub, the pubsub has been initialized.
3. Next, binding will be initialized in main thread. At this moment, it will use `h.SharedPubSubProvider` which is created and initialized at step 1 or use `CreateProvider` to create a new initialized pubsub.

So you can see, there's no case, v.vendor.RunLoop will use an uninitialized pubsub. No matter it is using `h.SharedPubSubProvider` or its own pubsub, it will be created and initialized in `CreateProvider` function.

Meanwhile, I believe @iwangjintian 's change doesn't change the existing flow. If there's really a case we need to improve, can we open an issue to resolve it separately? Then we can keep this PR simple.